Add recorder vars db_max_retries and db_retry_wait

TechnicallyJoe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Essentially, i'm proposing to add:
* db_retry_wait: <int>
* db_max_retries: <int>

to the recorder component. These are necessary because currently the retry count and sleep time is fixed to 10 retries, and 3 seconds sleep. I want to change that, so that its a variable.

The need came from an issue i had where my mariadb were consistently so slow to start, that my Home assistant container made its 10 retries and moved on. Testing shows that a completely new mariadb is about 2 mins to start up, so i was trying to figure out a way to force HA to check a bit more.
Proposed fixed were scripting the deploy, wrapping Home Assistant in a wait-for-it or dockerize container and using that to wait, but this seems like the best solution that benefits most people.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

default_config:

group: !include groups.yaml
automation: !include automations.yaml
script: !include scripts.yaml
scene: !include scenes.yaml

recorder:
  db_url: mysql://hass:hass@db:3306/hass?charset=utf8
  db_retry_wait: 10 # Wait 10 seconds between tries.
  db_max_retries: 20 # Try to connect 20 times.

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

I have not submitted an issue for this however we did have a lengthy dicussion about this on discord Installation>Core around 12:27 AM the 6th of february. Here we discussed possible other solutions and came to the conclusion that this was probably the best solution.

Also, disclaimer. This is my first run with python. I'm a capable developer otherwise, but this is the first time i've ever done anything with python. Please let me know if you got any comments or anything im missing.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

TechnicallyJoe committed:
added recorder vars db_max_retries and db_retry_wait

homeassistant commented:
Hi @TechnicallyJoe, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31561).

Once you do that we will be able to review and accept this pull request.

Thanks!

TechnicallyJoe committed:
fixed test_recorder_setup_failure

I failed because it was missing the two new variables. I simply added these with default values.

TechnicallyJoe committed:
fixed syntax error in test_recorder_setup_failure

TechnicallyJoe committed:
fixed formatting error in test_init_py for recorder component

TechnicallyJoe committed:
fixed typo in test case

springstan commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -96,6 +98,8 @@
                 ),
                 vol.Optional(CONF_DB_URL): cv.string,
+                vol.Optional(CONF_DB_MAX_RETRIES): cv.positive_int,
+                vol.Optional(CONF_DB_RETRY_WAIT): cv.positive_int,
```
Use the declared constants DEFAULT_DB_MAX_RETRIES, DEFAULT_DB_RETRY_WAIT as default values in the config schema. See the [official documentation](http://alecthomas.github.io/voluptuous/docs/_build/html/voluptuous.html?highlight=default%20value#voluptuous.schema_builder.Optional) for reference.
```suggestion
                vol.Optional(CONF_DB_RETRY_WAIT, default=DEFAULT_DB_RETRY_WAIT): cv.positive_int,
```

TechnicallyJoe replied:
Implemented suggested change in 8bac36bdd014bd9f2f5299aae40019bf23ea6ad7

springstan commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -138,13 +142,23 @@ def run_information(hass, point_in_time: Optional[datetime] = None):
+
+    db_retry_wait = conf.get(CONF_DB_RETRY_WAIT, None)
+    if not db_retry_wait:
+        db_retry_wait = DEFAULT_DB_RETRY_WAIT
```
Remove the if statements because they are not needed if a default value is set.

TechnicallyJoe replied:
Implemented the suggested change in 8bac36bdd014bd9f2f5299aae40019bf23ea6ad7

springstan commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -138,13 +142,23 @@ def run_information(hass, point_in_time: Optional[datetime] = None):
     if not db_url:
         db_url = DEFAULT_URL.format(hass_config_path=hass.config.path(DEFAULT_DB_FILE))
 
+    db_max_retries = conf.get(CONF_DB_MAX_RETRIES, None)
```
Use `dict[key]` for optional config keys that also have a default value set.

TechnicallyJoe replied:
I believe i did what youre asking in 8bac36bdd014bd9f2f5299aae40019bf23ea6ad7 - I'm not entirely sure. If you could review my solution, that would be great :-)

springstan replied:
Yes, thank you very much 👍  just one thing needs to be solved, so please take a look at my suggestion :)

TechnicallyJoe replied:
Ah! Gotcha. I've updated __init__.py with the changes you suggested in d11d446a7b42f8c21c8fe09329caa1b488109a12

TechnicallyJoe committed:
Updated the way the default keys for db_,max_wait and db_retry_wait is set

Implemented based on suggestions from @springstan

TechnicallyJoe committed:
Updated config_schema call to adhere to Black

springstan commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -138,13 +146,18 @@ def run_information(hass, point_in_time: Optional[datetime] = None):
     if not db_url:
         db_url = DEFAULT_URL.format(hass_config_path=hass.config.path(DEFAULT_DB_FILE))
 
+    db_max_retries = conf.get(CONF_DB_MAX_RETRIES)
```
Please use dict[key] here, e.g.
```suggestion
    db_max_retries = conf[CONF_DB_MAX_RETRIES]
```

TechnicallyJoe replied:
I've updated __init__.py with the changes you suggested in d11d446a7b42f8c21c8fe09329caa1b488109a12

springstan commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -138,13 +146,18 @@ def run_information(hass, point_in_time: Optional[datetime] = None):
         db_url = DEFAULT_URL.format(hass_config_path=hass.config.path(DEFAULT_DB_FILE))
 
+    db_max_retries = conf.get(CONF_DB_MAX_RETRIES)
+    db_retry_wait = conf.get(CONF_DB_RETRY_WAIT)
```
Same as above.

TechnicallyJoe replied:
I've updated __init__.py with the changes you suggested in d11d446a7b42f8c21c8fe09329caa1b488109a12

TechnicallyJoe committed:
changed conf.get to conf[dict] for var retrieval

springstan commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -133,18 +141,24 @@ def run_information(hass, point_in_time: Optional[datetime] = None):
         db_url = DEFAULT_URL.format(hass_config_path=hass.config.path(DEFAULT_DB_FILE))
 
+
+
```
Please remove these empty lines too.

TechnicallyJoe replied:
They have now been removed in 8112a1eadef2a658b2cca1f0313c07b22b199542 . Sorry. I'm having a hard time testing the code locally.

TechnicallyJoe committed:
removed 2 blank lines

springstan commented:
@TechnicallyJoe thanks for implementing my suggested changes 😊
Please update the documentation accordingly in the [documentation repo](https://github.com/home-assistant/home-assistant.io).

TechnicallyJoe commented:
@springstan 
https://github.com/home-assistant/home-assistant.io/pull/12023
Documentation PR added.

springstan commented:
Thanks @TechnicallyJoe!

springstan reviewed (approved):
LGTM 🎉
